### PR TITLE
Do not CSE constant operands for `GT_SIMD` vector constructors

### PR DESCRIPTION
**Pull request by TIHan**
**Description**
If a `GT_SIMD` op contains all constants and is an `Init`/`InitN`, then mark each constant as `DoNotCSE`. This logic already exists for `GT_HWINTRINSIC` nodes when creating vectors.

**Acceptance Criteria**
- [x] CI passes

[Diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1840616&view=ms.vss-build-web.run-extensions-tab)

### CODE REVIEW

**Comment by TIHan** [comment]
@dotnet/jit-contrib This is ready. Just will wait for CI to pass.

The diffs that I got were some improvements, but this is going to help us tweak the double constant heuristic in https://github.com/dotnet/runtime/pull/65176 since we do not want to hoist or CSE these kinds of constants.

**Comment by EgorBo** [comment]
There is already a similar logic in that method that skips CSE for IMM-able arguments and also for `Vector128/256.Create()` (because if we do CSE for any argument of constant Vector128 we likely won't be able to replace it with a data section load)

Can you show some examples? I assume you target `new Vector(...)` - in this case you can just extend the existing logic https://github.com/dotnet/runtime/blob/df3bd5502d2446a37a30f9064fa4f08b15c83f30/src/coreclr/jit/morph.cpp#L13923-L13938

**Comment by TIHan** [comment]
An example of this is:
```csharp
new Vector3(1, 0, 0)
```

I did see the existing logic there. Wanted to keep it separate but makes sense to combine them since they are so similar.

File: src/coreclr/jit/morph.cpp [hunk 1]
```diff
@@ -13912,6 +13912,29 @@ GenTree* Compiler::fgMorphMultiOp(GenTreeMultiOp* multiOp)
     }
 #endif
 
+#if defined(FEATURE_SIMD)
+    if (opts.OptimizationEnabled() && multiOp->OperIs(GT_SIMD))
+    {
+        bool allArgsAreConst = true;
+        for (GenTree* arg : multiOp->Operands())
+        {
+            if (!arg->OperIsConst())
```
Comment (EgorBo) [inline src/coreclr/jit/morph.cpp:13921]: btw, I assume OperIsConst is also true for GenTreeVecCon? in that case constant vectors as arguments also won't be CSEd? (or GenTreeVecCon only for GT_HWINTRINSIC?)

Comment (TIHan) [reply src/coreclr/jit/morph.cpp:13884]: I guess this was under the assumption that `Vector.Create` methods do not have any vector parameter types?

Comment (EgorBo) [reply src/coreclr/jit/morph.cpp:13884]: Yes, but your code doesn't check that it's Vector.Create or I am not reading it correctly?

Comment (TIHan) [reply src/coreclr/jit/morph.cpp:13884]: I see. I assumed `GT_SIMD` was only creation of vectors - that's not the case. I need to check for the constructors.

File: src/coreclr/jit/gentree.h [hunk 2]
```diff
@@ -8313,6 +8314,41 @@ inline bool GenTree::IsVectorZero() const
...
+                return true;
+
+            default:
+                return false;
+        }
+    }
+#endif // FEATURE_HW_INTRINSICS
+
+#ifdef FEATURE_SIMD
+    return OperIs(GT_SIMD);
```
Comment (tannergooding) [inline src/coreclr/jit/gentree.h:8346]: You probably want to check the ID is `SimdIntrinsicInit`, `...InitArray`, `...InitArrayX`, `...InitN`, or `...InitFixed`: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/simdintrinsiclist.h

File: src/coreclr/jit/morph.cpp (hunk 1 above)
Comment (tannergooding) [reply src/coreclr/jit/morph.cpp:13884]: Yes, but not for `Create`.

The check above of
> if (opts.OptimizationEnabled() && multiOp->IsVectorCreate())

should ensure that this is one of the Create calls that only takes primitive operands.

File: src/coreclr/jit/gentree.h (hunk 2 above)
Comment (tannergooding) [reply src/coreclr/jit/gentree.h:8346]: -- Looking at how this is used below, you probably want `Init` and `InitN`

`InitArray` and `InitArrayX` take a primitive array and so won't qualify the constant checks

Comment (tannergooding) [reply src/coreclr/jit/gentree.h:8346]: `InitFixed` might be interesting in that `new Vector3(new Vector2(5, 4), 3)` could qualify for transform to a `GT_CNS_VEC { 5, 4, 3 }`. But I expect that's not important for reducing regressions.

Comment (TIHan) [reply src/coreclr/jit/gentree.h]: I'm just going to look at `Init` and `InitN`.

**Comment by TIHan** [comment]
[Diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1840616&view=ms.vss-build-web.run-extensions-tab)

**Comment by TIHan** [comment]
@dotnet/jit-contrib This is ready again, diffs look pretty good. x64 has more improvements because we always hoist double constants in x64.

**Review by tannergooding — APPROVED** [review]

**Review by kunalspathak — APPROVED** [review]
LGTM

**Pull request merged**